Add Object Detection Tutorial

iizukak opened the pull request:
issue: #20

Adding object detection tutorial using Open Images V4 dataset.

iizukak commented:
I removed [WIP]. Need some review.

ruimashita commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,128 @@
+
+## Train a neural network
+
+You can do it like 
```
Why you not write the `$ ./blueoil.sh train config/{Model name}.yml` command here? Do you hesitate duplicate?
Also I think It is better the link is `./image_cls.html#train-a-neural-network` more than `./image_cls.html`.

ruimashita commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,128 @@
+
+## Convert training result to FPGA ready format.
+
+You can also do it like  <a href="./image_cls.html">Classification example</a>.
```
Also, here is not command, and the link don't have anchor.

lm-konda commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,128 @@
+
+## Preparation
+
+The Open Images V4 dataset is available from official website, but original dataset is too large for the tutorial. We provide reduced dataset.
```
I think we need "So". Like below.
So, we provide reduced dataset.

lm-konda commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,128 @@
+
+
+	$ wget https://s3-ap-northeast-1.amazonaws.com/leapmind-public-storage/datasets/openimages.tgz
+	$ tar xzf openimages.tgz
```
We can write as `tar xf` now. We don't need z.

lm-konda commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,128 @@
+
+    $ ./blueoil.sh init
+
+This is an example of configuration.
```
Command is "./blueoil.sh init". But there are no words like "initialize". 
In the first place, "init" may not be appropriate.

lm-konda replied:
For example, we write like below.
This is an example of initialization (or add some modifier).

lm-konda commented on docs/tutorial/image_det.md:
````diff
@@ -1,3 +1,128 @@
+apply quantization at the first layer?  yes
+```
+
+## Train a neural network
````
For consistency, we should unite train network / train network model.

lm-jira commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,128 @@
+
+## Run inference script on x86 Linux (Ubuntu 16.04)
+
+- Prepare inference images (not included in the training dataset)
```
In line 69, it maybe better to use {Model name} instead of [Model name].

iizukak commented on docs/tutorial/image_det.md:
````diff
@@ -1,3 +1,128 @@
+	      -c ../models/meta.yaml
+	```
+
+	*Tips:* The default thredhold for object detection is `0.05`. If you find too many boxses when running demo, you can edit `meta.yml` and set threshold about `0.4` or `0.5` as below code.
````
@lm-konda 
Thank you for comments.
Do you think how much threshold is appropriate?

lm-jira commented on docs/tutorial/image_det.md:
````diff
@@ -1,3 +1,166 @@
+	      -c ../models/meta.yaml
+	```
+
+	*Tips:* The default thredhold for object detection is `0.05`. If you find too many boxses when running demo, you can edit `meta.yml` and set threshold about `0.4` or `0.5` as below code.
````
boxses -> boxes

I think it maybe better to use
"set threshold to 0.4 or 0.5" if you want the uses to set the threshold to either 0.4 tor 0.5 or
"set threshold between 0.4 and 0.5" if any number in the range is okay.

ruimashita commented:
@iizukak Is it ready to merge ?

iizukak commented:
@ruimashita Sorry, not yet. I'll replace dataset.

lm-jira commented:
I wrote the report about hyper parameter tuning [here](https://leapmind.kibe.la/@jira/536).
I found that the optimal learning rate is `0.1` and should be scheduled by `tune2` method.
Moreover, the performance can also increase by increases batch size and epochs, but it results in longer training time.

ruimashita commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,169 @@
+	$ tar xf openimages.tgz
+
+
+This dataset consists of 5,000 color images in 2 classes, with 2,500 images per class.
```
It is friendly to write description what class are include, like this.
```
This dataset consists of 5,000 color images in images in 2 classes `Human face` and `Human hand`, with 2,500 images per class.
```

Don't you public the script to make subset from all open image v4?

iizukak replied:
@ruimashita 
The script need whole of the dataset. It's not useful for tutorial user.
And that script is not ophisticated enough to link...

iizukak replied:
Suggested sentence is fixed.

ruimashita replied:
@iizukak 
> The script need whole of the dataset. It's not useful for tutorial user.
And that script is not ophisticated enough to link...

All right, Thanks !

iizukak commented:
@lm-konda 
I fixed your commented issues.
Can you approve if looks good?

ruimashita commented:
> @lm-konda
> I fixed your commented issues.
> Can you approve if looks good?

@lm-konda 🉑 ?

lm-konda commented:
@iizukak @ruimashita I think OK!

ruimashita commented on docs/tutorial/image_det.md:
```diff
@@ -1,3 +1,169 @@
+
+When training is started, training log and checkpoints are generated under `./saved/{Mode name}_{TIMESTAMP}` directory.
+
+Training is running on TensorFlow backend. So you can use TensorBoard to visualize your training process. 
```
In order to user realize when the training finished, It is better to write how much MAP(Mean Average Presicion) are archived.

iizukak replied:
I agree.

iizukak commented:
@ruimashita 
I added loss, metrics and final detect boxes screenshot images of Tensorboard.
And updated `blueoil.sh init` parameter because of master brach updated.

Please check
https://deploy-preview-33--blueoil.netlify.com/tutorial/image_det.html

ruimashita commented on docs/tutorial/image_det.md:
```diff
@@ -16,18 +16,18 @@ Blueoil supports 2 formats for object detection.
 
 
-This dataset consists of 5,000 color images in 2 classes, with 2,500 images per class.
+This dataset consists of 2869 Human Face images and 5171 annotation boxes.
```
I think number of images and boxes is wrong.

I download openimages_face.tgz and extract it.

```
$ find ./images/ -type f -name "*.jpg" | wc -l
    2866
```
The number of images is `2866`.


```
$ cat annotations-bbox.csv | wc -l
    5171
```
`annotations-bbox.csv` is including header, so the number boxes is `5170 (5171 - 1)`

iizukak replied:
Thanks. I fixed.
